ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedColumnReader.java

```diff
@@ -0,0 +1,668 @@
+  }
+
+  void readBatch(
+    int total,
+    ColumnVector column,
+    TypeInfo columnType) throws IOException {
+
+    int rowId = 0;
+    while (total > 0) {
+      // Compute the number of values we want to read in this page.
+      consume();
+      int leftInPage = (int) (endOfPageValueCount - valuesRead);
+      int num = Math.min(total, leftInPage);
+      if (isCurrentPageDictionaryEncoded) {
+        LongColumnVector dictionaryIds = new LongColumnVector();
+        // Read and decode dictionary ids.
+        readIntegers(num, dictionaryIds, rowId, maxDefLevel,
+          (VectorizedValuesReader) dataColumn);
+        decodeDictionaryIds(rowId, num, column, dictionaryIds);
+      } else {
+        // assign values in vector
+        PrimitiveTypeInfo primitiveColumnType = (PrimitiveTypeInfo) columnType;
+        switch (primitiveColumnType.getPrimitiveCategory()) {
+        case INT:
+        case BYTE:
+        case SHORT:
+          readIntegers(num, (LongColumnVector) column, rowId, maxDefLevel,
+            (VectorizedValuesReader) dataColumn);
+          break;
+        case DATE:
+        case INTERVAL_YEAR_MONTH:
+        case LONG:
+          readLongs(num, (LongColumnVector) column, rowId, maxDefLevel,
+            (VectorizedValuesReader) dataColumn);
+          break;
+        case BOOLEAN:
+          readBooleans(num, (LongColumnVector) column, rowId, maxDefLevel,
+            (VectorizedValuesReader) dataColumn);
+          break;
+        case DOUBLE:
+          readDoubles(num, (DoubleColumnVector) column, rowId, maxDefLevel,
+            (VectorizedValuesReader) dataColumn);
+          break;
+        case BINARY:
+        case STRING:
+        case CHAR:
+        case VARCHAR:
+          readBinarys(num, (BytesColumnVector) column, rowId, maxDefLevel,
+            (VectorizedValuesReader) dataColumn);
+          break;
+        case FLOAT:
+          readFloats(num, (DoubleColumnVector) column, rowId, maxDefLevel,
+            (VectorizedValuesReader) dataColumn);
+          break;
+        case DECIMAL:
+          readDecimal(num, (DecimalColumnVector) column, rowId, maxDefLevel,
+            (VectorizedValuesReader) dataColumn);
+        case INTERVAL_DAY_TIME:
+        case TIMESTAMP:
+          break;
+        default:
+          throw new IOException("Unsupported");
+        }
+      }
+      rowId += num;
+      total -= num;
+    }
+  }
+
+  private void consume() throws IOException {
+    int leftInPage = (int) (endOfPageValueCount - valuesRead);
+    if (leftInPage == 0) {
+      readPage();
+    }
+  }
+
+  public void readIntegers(
+    int total,
+    LongColumnVector c,
+    int rowId,
+    int level,
+    VectorizedValuesReader data) throws IOException {
+    int left = total;
+    while (left > 0) {
+      consume();
+      readRepetitionAndDefinitionLevels();
+      if (definitionLevel >= level) {
+        c.vector[rowId] = data.readInteger();
+        c.isNull[rowId] = false;
+        if(c.isRepeating){
```

Comment:
not sure about this, but might be better to replace this with `c.isRepeating = c.isRepeating && (c.vector[0] == c.vector[rowId])` is better since the former has branching cost.

Also, space after ')'